spring-integration-core/src/main/java/org/springframework/integration/support/json/Jackson3JsonObjectMapper.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2025-present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.support.json;
+
+import java.io.File;
+import java.io.InputStream;
+import java.io.Reader;
+import java.io.Writer;
+import java.lang.reflect.Type;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import tools.jackson.core.JacksonException;
+import tools.jackson.core.JsonParser;
+import tools.jackson.databind.JacksonModule;
+import tools.jackson.databind.JavaType;
+import tools.jackson.databind.JsonNode;
+import tools.jackson.databind.ObjectMapper;
+import tools.jackson.databind.json.JsonMapper;
+
+import org.springframework.integration.mapping.support.JsonHeaders;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+
+/**
+ * Jackson 3 JSON-processor (@link https://github.com/FasterXML)
+ * {@linkplain JsonObjectMapper} implementation.
+ * Delegates {@link #toJson} and {@link #fromJson}
+ * to the {@linkplain ObjectMapper}
+ * <p>
+ * It customizes Jackson's default properties with the following ones:
+ * <ul>
+ * <li>The well-known modules are registered through the classpath scan</li>
+ * </ul>
+ *
+ * See {@code tools.jackson.databind.json.JsonMapper.builder} for more information.
+ *
+ * @author Jooyoung Pyoung
+ *
+ * @since 7.0
+ *
+ */
+public class Jackson3JsonObjectMapper extends AbstractJacksonJsonObjectMapper<JsonNode, JsonParser, JavaType> {
+
+	private static final boolean JODA_MODULE_PRESENT =
+			ClassUtils.isPresent("tools.jackson.datatype.joda.JodaModule", null);
+
+	private static final boolean KOTLIN_MODULE_PRESENT =
+			ClassUtils.isPresent("kotlin.Unit", null) &&
+					ClassUtils.isPresent("tools.jackson.module.kotlin.KotlinModule", null);
+
+	private final ObjectMapper objectMapper;
+
+	public Jackson3JsonObjectMapper() {
+		List<JacksonModule> jacksonModules = collectWellKnownModulesIfAvailable();
+		this.objectMapper = JsonMapper.builder()
+				.addModules(jacksonModules)
+				.build();
```

Comment:
I see Spring Framework does this instead:
```
 JsonMapper.builder().findAndAddModules(Jackson3JsonObjectMapper.class.getClassLoader()).build();
```
Maybe we check if really those well-known modules are picked up properly by Jackson 3 native API?
I mean this our new class might be a bit lighter without those inner classes and `collectWellKnownModulesIfAvailable()` ?

Reply:
By test results, it works as we expected.

But I think `JacksonJsonObjectMapper.class.getClassLoader()` would be the `ApplicationClassLoader`. This works because our Jackson dependencies only have these two modules, but if a user's application has other Jackson modules beyond Spring Integration... the situation might be different.

I'm a bit worried about this case. WDYT?

Reply:
> our Jackson dependencies only have these two modules

Those are optional in our case, so end-user has to add even these two in their project explicitly.
I'm not sure what problem you see since in the end any custom Jackson module has to be added into final product classpath to make them available.

Do you see something off with our ClassLoader arrangement?

We probably can use `ClassUtils.getDefaultClassLoader()` instead.
I always had a problem with understand how all those class loaders work 😢

Reply:
Ah, I understood that in the existing `Jackson2JsonObjectMapper`, no matter how many Jackson dependencies a user's product has, when created with the default ctor, only `wellKnownModules()` would be added.

This time, since it's changed to find modules from the loader, I thought that in such situations, when created with the default ctor, an `ObjectMapper` with additional `JacksonModules` beyond `wellKnownModules()` could be created.

I might be misunderstanding something though - please let me know if I'm thinking about this incorrectly! 🥲

Reply:
> only wellKnownModules() would be added.

Correct. That's the current behavior.

> with additional `JacksonModules` beyond `wellKnownModules()` could be created.

That's correct, too. Those "wellKnownModules" are just placeholder in our code to simplify class loading.
With this `findAndAddModules()` we definitely are going to load everything what would fit Jackson expectations, essentially whatever is there on classpath.

Having this discussion with you I have to go and double check a new code from Jackson.
Might be the case that this `findAndAddModules()` is not OK by default and we should comeback to `wellKnownModules()` instead.
Again: team consultation :wink:

Reply:
Thank you ! Actually I didn't know that.. Sorry for making you overthink this. 😞